tests/unit_tests/transformer/test_sparse_attention.py

```diff
@@ -0,0 +1,1202 @@
+            # Sparse attention specific configs
+            index_n_heads=8,
+            index_head_dim=64,
+            index_topk=self.index_topk,
+            rope_type='rope',
+            rotary_base=10000,
+            rotary_percent=1.0,
+        )
+
+        # Create indexer submodules spec
+        from megatron.core.extensions.transformer_engine import TELinear, TENorm
+        from megatron.core.transformer.spec_utils import ModuleSpec
+
+        indexer_submodules = IndexerSubmodules(
+            linear_wq_b=ModuleSpec(module=TELinear),
+            linear_wk=ModuleSpec(module=TELinear),
+            k_norm=ModuleSpec(module=TENorm),
+            linear_weights_proj=ModuleSpec(module=TELinear),
+        )
+
+        self.pg_collection = ProcessGroupCollection.use_mpu_process_groups(
+            required_pgs=['tp', 'cp']
+        )
+        self.indexer = Indexer(self.config, indexer_submodules, self.pg_collection)
+
+        yield
+        Utils.destroy_model_parallel()
+
+    def test_indexer_constructor(self):
+        """Test indexer initialization."""
+        assert isinstance(self.indexer, Indexer)
+        assert self.indexer.hidden_size == 256
+        assert self.indexer.index_n_heads == 8
+        assert self.indexer.index_head_dim == 64
+        assert self.indexer.index_topk == 32
+
+    @pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
+    def test_indexer_forward(self):
+        """Test indexer forward pass."""
+        seq_len = 64
+        batch_size = 2
+
+        self.indexer.cuda()
+
+        # Create input tensors
+        x = torch.randn(seq_len, batch_size, self.config.hidden_size, dtype=torch.bfloat16).cuda()
+        qr = torch.randn(seq_len, batch_size, self.config.q_lora_rank, dtype=torch.bfloat16).cuda()
+
+        # Forward pass
+        topk_indices = self.indexer(x, qr)
+
+        # Check output shape
+        assert topk_indices.shape == (batch_size, seq_len, min(self.config.index_topk, seq_len))
+        assert topk_indices.dtype == torch.long
+
+    @pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
+    def test_indexer_forward_with_scores(self):
+        """Test indexer forward pass with scores."""
+        seq_len = 16
+        batch_size = 2
+
+        self.indexer.cuda()
+
+        # Create input tensors
+        x = torch.randn(seq_len, batch_size, self.config.hidden_size, dtype=torch.bfloat16).cuda()
+        qr = torch.randn(seq_len, batch_size, self.config.q_lora_rank, dtype=torch.bfloat16).cuda()
+
+        # Forward pass with scores
+        index_scores, topk_indices = self.indexer.forward_with_scores(x, qr)
+
+        # Check output shapes
+        assert index_scores.shape == (batch_size, seq_len, seq_len)
+        assert topk_indices.shape == (batch_size, seq_len, min(self.config.index_topk, seq_len))
+        assert index_scores.dtype == torch.float32
+        assert topk_indices.dtype == torch.long
```

Comment:
Please also check the indeces' range and uniqueness
```suggestion
        assert topk_indices.dtype == torch.long
        assert torch.all((topk_indices >= 0) & (topk_indices < seq_len))
        assert torch.all(
            torch.sort(topk_indices, dim=1).values[:, 1:] != torch.sort(topk_indices, dim=1).values[:, :-1]
        )
```

Reply:
Added.